modules/packages/ZMQ.chpl

```diff
@@ -1024,11 +1023,16 @@ module ZMQ {
         if (0 != zmq_msg_close(msg)) {
           try throw_socket_error(errno, "recv");
         }
+        return val;
+      }
 
-        // Return the value to the calling locale
-        ret = val;
+      if here == classRef.home {
+        return innerRecv();
+      } else {
+        var localRet: T;
+        on classRef.home do localRet = innerRecv();
+        return localRet;
```

Comment:
I believe we do similar things elsewhere. I wonder what do we need to improve to avoid these. Is it just not being able to return from `on` statements that forces us into this idiom, or is there something more to it? Do you have any idea?

Reply:
Yeah, I think it's just not being able to return from an on-stmt that forces this idiom.

Reply:
See #13587 and #15808 which are different strategies for solving the same problem.